minicopier-master/src/minicopier/TransferTask.java

```diff
@@ -0,0 +1,185 @@
+/*
+  TransferTask.java / MiniCopier
+  Copyright (C) 2007-2009 Adrian Courrèges
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License as
+  published by the Free Software Foundation; either version 2 of
+  the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+*/
+
+package minicopier;
+import java.io.*;
+
+
+public class TransferTask extends Thread{
+
+	private Copier copier;
+	
+	private File source;
+	
+	private File destination;
+	
+	private File destinationFolder;
+	
+	private Boolean append;
+	
+	private Boolean paused;
+	
+	private Boolean cancel;
+	
+	private long totalSize;
+	
+	private long currentSize;
+	
+	boolean completed;
+	
+	boolean putInQueue;
+	
+
+	
+	public TransferTask(Copier c, FileToTransfer file, boolean _app){
+		this.copier = c;
+		this.source=file.getSourceFile();
+		this.destination=file.getDestinationFile();
+		this.destinationFolder=file.getDestinationFolder();
+		this.append=_app;
+		this.paused = false;
+		this.cancel = false;
+		this.totalSize = file.getSize();
+		this.currentSize = 0;
+		this.completed = false;
+		this.putInQueue = false;
+	}
+	
+	
+	public boolean getCompleted(){
+		return this.completed;
+	}
+	
+	
+	public void setPause() {
+		this.paused = true;
+	}
+	
+	public void unpause() {
+		synchronized(this) {
+		this.paused = false;
+		this.notifyAll();
+		}
+	}
+	
+	public void setCancelAndQueue(boolean _putInQueue){
+		this.cancel = true;
+		this.putInQueue = _putInQueue;
+	}
+	
+	public boolean getPutInQueue(){
+		return this.putInQueue;
+	}
+	
+	public boolean getCancel() {
+		return this.cancel;
+	}
+	
+	public void run() {
+		// Streams declarations
+		ThroughputLimitInputStream sourceFile = null;
+        java.io.FileOutputStream destinationFile = null;
+        
+        // Read by segment of 0.5Mo 
+        int bufferSize = 512*1024;
+        
+        this.copier.setFileTotalSize(this.totalSize);
+        this.copier.setFileCurrentSize(0);
+        
+        try {
+        		//creation of directories (if needed)
+        		this.destinationFolder.mkdirs();
+                // Creation of file (nothing done if already present)
+                this.destination.createNewFile();
+                
+                // Opening streams
+                sourceFile = new ThroughputLimitInputStream(new java.io.FileInputStream(source));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by FileOutputStream constructor. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).